### ISSUE TITLE
Help wanted. No NIF was found.

### ISSUE THREAD

**Issue by kaaquist** [kaitoy/pcap4j#174]
I have run pcap4j on RaspBerryPI3 and everything is great. Now I run it on another small computer board and I get this error: 
```
2018-06-27 19:19:05 [main] 822  INFO  org.pcap4j.core.NativeMappings - Pcap4J successfully loaded a native pcap library: libpcap version 1.7.4 
2018-06-27 19:19:06 [main] 1496 INFO  org.pcap4j.core.Pcaps - No NIF was found.
```
There is a libpcap-dev installed and I can run `tcpdump` on the interface I would like to listen in on. But I keep getting this error. 
I do start my code as `root`. 
OS is Ubuntu 16.04 
Libpcap v. 1.7.4 
Arm64 Processor. 

I have tried to reinstall all libpcap* packages and I can get the code ot fail if there are no libpcap installed. 
Ok, please help. 

Thanks.

**Comment by koolmano** [comment]
How exactly do you run pcap4j? Using tar? If you using tar you need to run it like sudo java -jar your.jar

**Comment by kaaquist** [comment]
Yes I use a jar, and I do run it as sudo. But would that be the error if I don't run it as sudo user?

**Comment by koolmano** [comment]
well, I tried to use without sudo it failed on my machine. I belive that libpcap use only super user mode, for example on my machine it is imposible to run tcpdump without sudo it tells this:
"tcpdump: eth2: You don't have permission to capture on that device
(socket: Operation not permitted)" And tcpdump uses libpcap likewise. So, without super user permission I didn't have any luck.

**Comment by kaaquist** [comment]
That is also my experience. Hmm Thanks for helping. 
I do use a sudo user, I'm just not aware if there are any sudo users that does not have the right rights? 
I'm creatin a small program that does just list the NIF/NIC's that libpcap/pcap4j can see.  
I will come back with the results.

**Comment by koolmano** [comment]
good luck, man. with this library you need this. :)

**Comment by kaitoy** [comment]
@kaaquist 
You may need to reboot your machine.
I've ever experienced libpcap didn't work just after installing it and started working after a reboot.

**Comment by kaaquist** [comment]
Thanks. I will try that too..

**Comment by kaaquist** [comment]
I have created this small test, (from sample code): 
```java
import org.pcap4j.core.*;
import org.pcap4j.util.NifSelector;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;


public class TestNic {
    private static final Logger logger = LoggerFactory.getLogger(TestNic.class);

    private static final String COUNT_KEY
            = TestNic.class.getName() + ".count";
    private static final int COUNT
            = Integer.getInteger(COUNT_KEY, 10);

    private static final String READ_TIMEOUT_KEY
            = TestNic.class.getName() + ".readTimeout";
    private static final int READ_TIMEOUT
            = Integer.getInteger(READ_TIMEOUT_KEY, 100); // [ms]

    private static final String SNAPLEN_KEY
            = TestNic.class.getName() + ".snaplen";
    private static final int SNAPLEN
            = Integer.getInteger(SNAPLEN_KEY, 65536); // [bytes]

    private static final String BUFFER_SIZE_KEY
            = TestNic.class.getName() + ".bufferSize";
    private static final int BUFFER_SIZE
            = Integer.getInteger(BUFFER_SIZE_KEY, 1024 * 1024); // [bytes]

    private static final String TIMESTAMP_PRECISION_NANO_KEY
            = TestNic.class.getName() + ".timestampPrecision.nano";
    private static final boolean TIMESTAMP_PRECISION_NANO
            = Boolean.getBoolean(TIMESTAMP_PRECISION_NANO_KEY);

    private static PcapHandle getPcapHandlerB(PcapNetworkInterface nif) throws PcapNativeException {
        PcapHandle.Builder phb = new PcapHandle
                .Builder(nif.getName())
                .snaplen(SNAPLEN)
                .promiscuousMode(PcapNetworkInterface.PromiscuousMode.PROMISCUOUS)
                .timeoutMillis(READ_TIMEOUT)
                .bufferSize(BUFFER_SIZE);
        if (TIMESTAMP_PRECISION_NANO) {
            phb.timestampPrecision(PcapHandle.TimestampPrecision.NANO);
        }
        return phb.build();
    }


    public static void main(String[] args) throws PcapNativeException, NotOpenException, IOException {
        String filter = args.length != 0 ? args[0] : "";

        System.out.println(COUNT_KEY + ": " + COUNT);
        System.out.println(READ_TIMEOUT_KEY + ": " + READ_TIMEOUT);
        System.out.println(SNAPLEN_KEY + ": " + SNAPLEN);
        System.out.println("\n");

        PcapNetworkInterface nif;
        try {
            nif = new NifSelector().selectNetworkInterface();
        } catch (IOException e) {
            e.printStackTrace();
            return;
        }

        if (nif == null) {
            System.out.println("Nothing found!");
            return;
        }

        System.out.println(nif.getName() + "(" + nif.getDescription() + ")");
    }
}
```
And that give me an error when I run it as sudo user: 
```bash
root@tegra-ubuntu:/home/ubuntu/test_nic# ./gradlew run -PclassToExecute=TestNic
:compileJava UP-TO-DATE
:processResources NO-SOURCE
:classes UP-TO-DATE
:run
TestNic.count: 10
TestNic.readTimeout: 100
TestNic.snaplen: 65536


13:23:18.736 [main] INFO org.pcap4j.core.NativeMappings - Pcap4J successfully loaded a native pcap library: libpcap version 1.7.4
13:23:19.400 [main] INFO org.pcap4j.core.Pcaps - No NIF was found.
java.io.IOException: No NIF to capture.
	at org.pcap4j.util.NifSelector.selectNetworkInterface(NifSelector.java:44)
	at TestNic.main(TestNic.java:69)

BUILD SUCCESSFUL in 2s
2 actionable tasks: 1 executed, 1 up-to-date
root@tegra-ubuntu:/home/ubuntu/test_nic# exit
exit
ubuntu@tegra-ubuntu:~/test_nic$ 
```
I will try to create a java class that can tell me what privileges the sudo user has when running as `sudo`. 
Thanks for helping.

**Comment by kaaquist** [comment]
Has anyone ever had to run the following command to get libpcap to work with java in ubuntu: `sudo usermod -a -G netdev <user>`, I have tried this on the setup I have. I did not work, but I was worth trying it out. 
This here is the OS: `Linux tegra-ubuntu 4.4.38-tegra #1 SMP PREEMPT Mon Apr 30 16:06:06 CEST 2018 aarch64 aarch64 aarch64 GNU/Linux`. 
I just done get why it does not work. Has anyone ever had that error that they can not run the list all devs command. Even without being sudo user?

**Comment by kaaquist** [comment]
Ok, I found that on the device I'm running my code there was no `libjna-java` installed and therefor I was not able to list NIFs. 
Now I get a new error: 
I can see all TCP traffic but the UDP traffic does not get captured. 
I have seen this error once before, where there was no installation of the `libpcap-dev`. But in this here case I do have that installed.. So has anyone encountered a similar problem? 
Thanks.

**Comment by kaaquist** [comment]
Yes, that is true. 😁 
I did  use strace and gdb. Gdb did not give me any output since the app does not crash, or alle exceptions are handled.  
But in the end, (today), I did find the problem. On the ARM board I did not have a `libjna-java` installed and linked to the JRE I'm using to execute the PCAP4J app, I guess that somewhere in `PCAP4J` an error form the Native call should have flagged this/or propagated the error. 
The only reason I found this was because I created a Java app that accessed `libpcap` in a different way, not using `PCAP4J`. 
But thanks for the help.
